designs/plugin-resolution.md

```diff
@@ -0,0 +1,88 @@
+# Plugin and project version resolution
+
+## Old behavior
+
+The same process is followed for all commands, without distinction between `init` and the rest. The only difference is
+that, for `init`, the configuration file will not be present.
+
+1. Obtain the project version from project configuration file's `version` field, `--project-verison` flag, or default
+   project version configured during CLI creation (`cli.New(cli.WithDefaultProjectVersion(...))`).
+    * In case both a project configuration file and the flag are found, fail if they are different.
+2. Obtain the plugin keys from the project configuration file's `layout` field, `--plugins` flag, or default plugins
+   configured during CLI creation for the above project version (`cli.New(cli.WithDefaultPlugins(...))`).
+    * In case both a project configuration file and the flag are found, fail if they are different.
+3. In case any of the plugin keys is not fully qualified (full name and version), check if there is only one plugin that
+   fits that unqualified plugin key.
+4. Resolve the plugins with the list of qualified keys above.
+5. Verify the plugins support the project version.
+   
+
+### Flaws
+
+* Default plugins need to be provided per project version
+* Project version can't be resolved (e.g., when a single project version is supported by the provided plugins).
+* Command calls are not allowed to override plugin keys.
+
+## Proposal
+
+The following plugin resolution algorithm solves the flaws described above:
+
+1. [All commands but Init] Obtain the project version and plugin keys from the project configuration if available.
+     * If they are available, plugin keys will be fully qualified.
+2. [Init command only] Obtain the project version from `--project-version` flag.
+     * The flag is optional, we will try to resolve project version later in case it is not provided.
```

Comment:
There will always be a default project version. If no default plugins support the default project version, then an error should be returned.

Reply:
The idea here is, in case they provide a project version, make decissions knowing it, but in case it wasn't provide, just make decissions as if it was unknown and try to resolve it later.

Imagine I have the following plugins resolved:
Plugin | Supported project version
---|---
go/v2 | 2, 3
go/v3 | 3
declarative/v1 | 2, 3

If the user provides a `--plugins=go/v3,declarative` flag, we would resolve to the last two plugins, and the only supported project version by both of them is `3` so we can implicitly resolve the project version.
If the user provides a `--plugins=go/v2,declarative` flag, we would resolve to the first and last plugin, and two different project versions are supported. If any of those is the default project version configured for the CLI, that will be used.

So we will have 3 ways of resolving the project version:
- Explicitly by providing `--project-version` flag or having an already present config file (`version` field).
- Implicitly by providing a set of plugins that only support a single common project version.
- Default in case all resolved plugins support the default project version defined in the CLI.